### ISSUE TITLE
Why does the trajectory tracking on a real machine feel jerky?

### ISSUE THREAD

**Issue by Zedzou** [HKPolyU-UAV/qr_mpc#5]
thank you for your code, I have successfully deployed on a real device, but why does the trajectory tracking on a real machine feel jerky? Do you have any ideas.

**Comment by RockyJBL** [comment]
Thank you for your feedback, and I'm glad to hear you've successfully deployed the code on a real device.

For real-world deployment, it could be caused by multiple reasons. For MPC side, you might try reducing the position weights in ``generate_c_code.py`` or tuning the hovering thrust parameter in ``quadrotor.py``. On the system side, please check the accuracy of position feedback, the tracking performance of the inner-loop controllers, and any potential transmission delays, as these can all impact smooth trajectory tracking.

**Comment by GavindotYang** [comment]
Thanks for sharing, if I want to deploy this on a real machine as well, is it suitable to use px4?

**Comment by RockyJBL** [comment]
> Thanks for sharing, if I want to deploy this on a real machine as well, is it suitable to use px4?

Yes, of course. We used PX4 based flight controllers in experimental validations.
